### DISCUSSION TITLE
OpenSceneGraph "new" version

### DISCUSSION THREAD

**Discussion by jurgus** [openscenegraph/OpenSceneGraph#1285] [Ideas]
Hi everyone, I wanted to introduce to you a mine personal hobby project to "improve" OSG. My "improvements" that I am working on can be summarized in the following points.
- Upgrade OSG to use only the latest version of OpenGL in the Core specification. And to "replace" the older OpenGL functionality with the appropriate "shaders" (that's still WIP). Along with this upgrade, I would like it to become possible to debug the OSG applications using, for example, RenderDoc (and using RenderDoc I was able to find a very subtle error in the osgdeferred sharder example code)..
- Modernization of C++ to the C++17 standard. With modern C++, I was able to abandon the dependence on the OpenThread library and base all "threading" code on what the C++ library gives programmers. As of now, this work has been completed and I have not found any bugs in the new code so far.
- Cleaning and optimization. OSG is very good at rendering "static" 3D worlds, but when "dynamic" elements appear (this is more apparent as these dynamic objects become more numerous), it shows that IMHO it was not designed to handle these dynamic objects at scale. This is also a WIP, but the first optimization in this area has been done.

Does this "new" version maintain backward compatibility with the latest OSG version available on Github? Unfortunately, NO. I know this is probably not good news. But changes to API (OSG 3.6.5) I try to keep to the "bare minimum". And they, at the moment, around the Math and Image/Texture classes and enforced by removed obsolete code.

What are the effects of the changes I made?

I'll show a more subtle change first, and that is full support for Linear-Space Lighting (needed for PBR rendering)

-Linear-Space Lighting
![SpaceLinear](https://github.com/openscenegraph/OpenSceneGraph/assets/7025169/56ffa7b2-6752-4127-85fd-bae89bbaeb8f)

-Gamma-Space Lighting
![SpaceSRGB](https://github.com/openscenegraph/OpenSceneGraph/assets/7025169/bb11d0b2-6ff6-47e6-81cd-9174cdd85d59)

Next I want to show performance improvements.
Test was made by running slightly modified osgdeffered example (just added more 4000 torus).

![OSG_4 0](https://github.com/openscenegraph/OpenSceneGraph/assets/7025169/f731f3e3-2dac-4847-abf4-a458a85e5923)

Lets compare it with stack version of OSG ( 3.6.5).

![OSG_DL](https://github.com/openscenegraph/OpenSceneGraph/assets/7025169/1bb03623-edca-4b72-834e-e3b54afb1752)

As we can see “optimization” do their work. We gain on every step except Draw. And this is mostly because by default OSG is using DisplayList for rendering. On PC they are still supported but on mobiles they not. 

Lets test OSG with disabled usage of DisplayLists and turn off it in Cmake.
 
![OSG_V_ARRAY](https://github.com/openscenegraph/OpenSceneGraph/assets/7025169/77ecf67a-d30c-41fe-bdf8-f9bde65b8e46)

and IMHO something rotten in this rendering path.

What's next.
- Ending the emulation of legacy OpenGL functions. This will enable the removal of unnecessary code from the performance-critical rendering path.
- Cull pass optimization on huge "dynamic" worlds.
- Completing the cleaning of the math source code. And I'm thinking of adding an explicit path for SSE, AVX math operations.
- Optimize Draw Pass (I have some ideas).
- Optimize Update Pass. I need to test whether swapping "some" of the work from cull to update is beneficial.
- Add support for SPIR-V shaders.
- Adding "programmer friendlier" creation of "render passes".
- Cleaning up the source code. And thinking about removing "obsolete" code.
- Adding a GLTF and USD importer.
- The sky is the limit.

To summarize this post. This is my beginning journey in optimizing and upgrading OSG. As I mentioned, this is my hobby project, but I look forward to your feedback, in particular,  for now I have only done a synthetic test of this "new" version of OSG. I expect that someone can share with me a more representative scene or a scene where I can conduct appropriate performance and comparability tests.

**Comment by robertosfield** [comment]
The OSG is legacy scene graph being maintained in a backwards compatible fashion for existing applications.  Breaking backwards compatibility looses this fundamental asset of the OSG for existing users.

If you want a modern take on the OSG then you need to move the VulkanSceneGraph, it's C++17, it uses Vulkan, it addresses all the major CPU performance bottlenecks of the OSG, it's far better designed for modern hardware and modern graphics/compute applications.

Back in 2018 I decided doing an upgrade like you have in mind was pointless, as it couldn't come close to delivering the full performance and feature capabilities that modern hardware can offer, it's why I chose to develop the VSG.  Now, 5 years on my view has only solidified behind creating the VSG as being the right choice for the scene graph users.  

Now, the OSG is open source, you are free to build something that isn't backwards compatible, still uses the legacy graphics API that now gets no updates, because Khronos is focused on Vulkan,  not OpenGL, because like the OSG, OpenGL is now just being maintained for legacy applications.

Or if you have time you want to invest in learning modern scene graphs and graphics/compute APIs then you could look towards seeing what value you can add to a the VSG project or the collection of projects that build upon it.

If you do decide the create a new "OSG" like project then may I ask you call it something other than the OSG to avoid creating lots of confusion amongst the community.

**Comment by robertosfield** [comment]
In case folks are wondering about items on @jurgus list.  The VulkanSceneGraph already has PBR support, with glTF loader that fully supports it and much more robust [cascaded shadow maps](https://github.com/vsg-dev/VulkanSceneGraph/discussions/991) support out of the box that goes well beyond what osgShadow is capable of.

This month we have also been working on providing users a richer experience with annotation the scene graph when using with RenderDoc, and also working on provide hooks for Tracy instrumentation.  Details on this VulkanSceneGraph discussion thread: 

https://github.com/vsg-dev/VulkanSceneGraph/discussions/1046

There are still gaps in functionality between the VSG and OSG, but bit by bit these are being filled in, and when they are they leap-frog what the OSG is capable of.

**Comment by jurgus** [comment]
Hi Robert, thank you for your reply.
When I talked about the "new" version of OSG, I had no bad intentions of creating confusion in the OSG community.
The name of the "project" I am working on is on the last place of my priorities. And I agree with you, and since 
I will soon try to make the source code public (as I better test it), I will try to make sure that the name cannot be confused with OSG.

P.S.
I know, projesional experience working in Computer Graphics I know that working with Vulkan is not a trivial task, 
so I wish all the best to the VSG library. If I can help in any way, I will not fail to do so.

**Reply by xarray** [reply]
Hi, it is cool to see someone else who want to work on a modern version of OSG.

Have a look at osgverse (https://github.com/xarray/osgverse). It already includes some features that you may be interested in, like an initial version of PBR shading and deferred pipeline, non-intrusive cascaded shadow maps and deferred lighting, GLTF & FBX importing (with full skeleton / blendshape supports), as well as WASM support for running OSG scene on web browsers. You may feel free to use my code to build your "new" version of OSG, or contribute to osgverse to fix its tons of bugs and finish the longer and longer TODO list. :)

**Comment by jurgus** [comment]
Hi Rui,
osgverse looks great.
I am an engineer/programmer focused on CPU/GPU performance. But if your library's CPU/GPU performance is unsatisfactory (mostly in a dynamic worlds where many objects move or are animated), you might be interested in building osgverse on the modernized, optimized version of OSG that I'm working on :).

**Reply by robertosfield** [reply]
If you care about performance on CPU and GPU you wouldn't be considering using the OSG and OpenGL as your starting place, both are outdated and have high CPU overhead.

I'm the lead author of the OSG,I used OpenGL every working day between 1992 to 2018, I am not so random developer talking shit about stuff I don't have any knowledge and experience of.

In the vsgTuorial online book I've been working on I touch upon the motivations for moving to a completely new scene graph and the go into the performance benefits of doing so:

https://vsg-dev.github.io/vsgTutorial/

You will never be able to achieve the performance and functionality that can be achieved with the VSGulkan/V, and if you are breaking backwards compatibility what is the point of accepting compromised performance from day one?

**Comment by jurgus** [comment]
Hi Robert,
I have a lot of experience working with various game engines. And I know from experience that switching from OpenGL to Vulkan (or from DirectX 11 to 12) is not a silver bullet for engine performance problems. 
Using the latest APis is not a trivial task, and doing it correctly is even more difficult :). It took a lot of work for the developers of Unreal Engine or Unity to make their engines on Vulkan or DirectX 12 run faster than on OpenGL or DirectX11.
There are many pitfalls in DX12 or Vulkan when using them, e.g. barriers, memory allocation, PSO, resource descriptor handling. Not all of these problems but a good portion are solved by driver engineers for the user in OpenGL. It's not a perfect solution, but they get the job done.

OpenGL and OSG will have little chance of surpassing VSG when it is ready. But what I want to achieve, at least give it a fighting chance :). 

I ask myself this question how much more potential there is in OSG and OpenGL, and this hobby project of mine is also to check it out.

**Reply by robertosfield** [reply]
The VSG has already solved the major challenges of working with Vulkan, many tasks are now actually easier with the VSG than they are with OSG/OpenGL.  A key goal of the VSG was to make Vulkan easy.

The main reason why graphics software fails to achieve the potential performance improvements promised by Vulkan is that they are poorly architected for them and have a CPU overhead that is so high that this dominants performance, not the graphics API.  Fix those problems are you start to see the benefit of Vulkan.  

The VSG is designed specifically around Vulkan and the capabilities of modern hardware.  Performance improvement you see between OSG/OpenGL to VSG/Vulkan is typically 3-4x faster, not a couple %.  You can lower the CPU overhead that the OSG has by adopting the techniques I've implemented in the VSG to try and close this gap, but you can't fix how OpenGL works, but this would fundamentally break backwards compatibility for existing OSG codebase and users.

Please read the vsgTutorial, spend some time learning the VSG, this time will be a invaluable in understanding the context that the OSG/OpenGL & VSG/Vulkan have in history through to  technical side to different design decisions made.  Even if folks don't actually use the VSG in the end this should still help them understand the topics of graphics and scene graphs better.

OpenGL is zombie at this point.  OpenGL 4.6 was released in July 2017, over 6 years ago.  It's only future is keeping existing OpenGL based applications working.  Khronos know this, and by now I would have hoped that the rest of the industry would have come to terms with this.

Contrast this to Vulkan, it's very much alive and active development by Khronos with lots of hardware companies support latest releases quickly, and the VulkanSDK keeps getting regular updates even between point releases of Vulkan itself.  It's a healthy and vibrant ecosystem.  

Vulkan already beats OpenGL on features - compute shaders, cross vendor mesh shaders and ray tracing, multi-thread design, explicit memory management, and as new hardware features are being developed they are being exposed in Vulkan, initially via extensions then pulled in to the core Vulkan.

All while the OpenGL horse which not technically dead is laying in Coma, unchanged for over 6 years.  This beast isn't suddenly going to get up and start running.

Of all the middle-ware graphics APIs available, the VSG is closest in design and philosophy to the OSG, it's born from the same DNA so for existing OSG while requiring a major port is probably the route with least pain.  My job is get the VSG as quickly to a place to where OSG users can port to it and know they have a solid foundation for the next 20 years.

If developers have time on their hands and energy to invest in something new then please consider contributing to the VSG project, there are plenty of areas that one could make their own - the VSG project is composed of a family of projects, rather than monolithic one like the OSG.

Trying to build something on top the existing OSG/OpenGL is flogging what is close to a dead horse, it bifurcates development communities and does nothing with pushing the start of art along.

**Comment by jurgus** [comment]
I am keeping my fingers crossed for the successful development of the VSG project. I believe that having a next-generation VSG and some improvements around OSG sounds like a win-win situation for everyone.

**Comment by robertosfield** [comment]
If developer resources were an infinite commodity then trying to pull in two different directions at once might be fine.  In the real world we are finite resource it's better if we all pull in the same direction.

For existing OSG users what will they decided to do?  Try porting to your non backwards compatible OSG version in the hope that it isn't too painful and not too much useful functionality is lost, or just bite the bullet and port to the VSG or other properly modern API.

Right now, my advice to existing OSG users is maintain your current software with OSG-3.6.5 and start learning VSG/Vulkan so when you get the opportunity you can write new applications based on this, or port your existing code base.

The % of OSG users that can port across to the VSG increases with each step forward with the VSG, but the more effort that happens to fill in the functional gaps missing from the VSG the quicker we can increase that % of OSG users that can find a safe haven in the VSG.

Having developers waste time on the deluded idea that OSG can somehow be tweaked to make it modern while they could be doing something productive within a project that has a long term and vibrant future is galling to me, rather than an dead-end that Khronos sealed 6 years ago when they stopped developing OpenGL.

However, if I can't help you see sense then you are free to waste your time on fools errand then I have to hope my skill at developing scene graph is better than my skills are persuasion.

**Comment by Mannshoch** [comment]
If I see it right the OpenSceneGraph is not anymore in development. 
Please update the Github Readme according this and may find a solution for the [Website](http://www.openscenegraph.org/)
Please be straight open and tell people what's the current case.

**Reply by robertosfield** [reply]
Please just read what I've been writing over the past 5 years.  Posts to this forum and to the website have all made it clear that the successor to the OpenSceneGraph, VulkanSceneGraph has been development and the OSG is in maintenance mode since 2018.  

My _2023 review and 2024 plans_ post goes into some further details: #1291.

We also have a new website in development: #1292 which makes clear that the OSG is maintenance mode and VSG should be used for new projects.

**Reply by v-fox** [reply]
So, people supposed to magically derive the most important thing about your project from some obscure links under 5 layers of sub-menus instead of title and main README? For me this whole thing is no more than a broken dependency for Scribus (which should probably just migrate to Qt3D). Why do I have to use search in pile of obscure forum-threads to known if this thing is even buildable?

You might as well write it in chalk on a wall of your apartment, photograph it and post description-less on some social media, so users would have to use AI-based OCR-capable web-search to find it, as some kind of novel clue-hunting game.
